### PR TITLE
Add associative scan

### PR DESCRIPTION
**Pull request by SamDuffield**
First attempt at using `jax.lax.associative_scan` #14 , but it's throwing a matmul contracting dimensions error and I'm not sure why.

### CODE REVIEW

File: thermox/sampler.py [hunk 1]
```diff
@@ -57,21 +57,49 @@ def transition_cov_sqrt_vp(v, dt):
...
     dts = jnp.diff(ts)
 
-    xs = scan(scan_body, (x0, key), dts)[1]
-    xs = jnp.concatenate([jnp.expand_dims(x0, axis=0), xs], axis=0)
-    return xs
+    noise_terms = jax.vmap(transition_cov_sqrt_vp)(
+        jax.random.normal(key, (len(dts),) + x0.shape), dts
+    )
+
+    def binary_associative_operator(elem_a, elem_b):
```
Comment (AdrienCorenflos) [inline thermox/sampler.py:76]: Just vmap this function.

**Review by AdrienCorenflos — COMMENTED** [review]
Associative scan takes vectorised operators (given that the operations across the leaves of the computational tree).

**Comment by SamDuffield** [comment]
Update: `associative_scan` now working but seems like something is wrong with the calculations so I need to check the maths again

**Comment by SamDuffield** [comment]
Ok I fixed the maths! At the cost of doubling the number of `expm_vp` calls, we might be able to halve it again with further thought although I'm not sure.

Next step is to add `associative_scan` for `log_prob`

File: tests/test_sampler.py
Comment (KaelanDt) [inline tests/test_sampler.py:1]: Would be good to add a test for the associative scan feature by itself (check the log(n) dependence in the number of workers). I'll take a stab at this

File: thermox/sampler.py [hunk 2]
```diff
@@ -27,6 +27,9 @@ def sample_identity_diffusion(
     Preprocessing (diagonalisation) costs O(d^3) and sampling costs O(T * d^2)
     where T=len(ts).
 
+    Uses jax.lax.associative_scan, so will run in time O(log(T) * d^2) on a GPU/TPU with
```
Comment (KaelanDt) [inline thermox/sampler.py:30]: My understanding is that this should also work on multiple cores/workers of a cpu. Also, in most cases if `T=len(ts)` it's rare that `T` cores are available, we could clarify this in the docstring

Comment (KaelanDt) [reply thermox/sampler.py:30]: We can add the complexity on page 2 here https://www.cs.cmu.edu/%7Eguyb/papers/Ble93.pdf

**Comment by KaelanDt** [comment]
Finished the speedup comparison with and without associative scan and adapted the handling of random keys in `_sample_identity_diffusion`, should be good to review

**Comment by SamDuffield** [comment]
Be sure to add the underscore to `sample_identity_diffusion` to become `_sample_identity_diffusion` and maybe remove the docstring too

**Review by KaelanDt — APPROVED** [review]
lgtm!

**Pull request merged**